### PR TITLE
IPv6/multi : lowered the complexity of FreeRTOS_UDP_IP.c

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
In this PR I worked on the complexity of the code in FreeRTOS_UDP_IP.c
It had only 2 functions: send and receive an UDP packet. I have split these functions up into a total of 10 functions.

Here are the new complexity scores for this module:
~~~
Score | ln-ct | nc-lns| file-name(line): proc-name
 1      20      16   FreeRTOS_UDP_IP.c(265): prvUDPSendPacket
 2      21      18   FreeRTOS_UDP_IP.c(117): prvSetIPHeaderForICMP
 2      34      26   FreeRTOS_UDP_IP.c(73): prvLookupIPInCache
 3      15      13   FreeRTOS_UDP_IP.c(290): prvFindIPv4Endpoint
 3      43      31   FreeRTOS_UDP_IP.c(148): prvSetIPHeaderForUDP
 4      50      33   FreeRTOS_UDP_IP.c(205): prvStartLookup
 4      52      43   FreeRTOS_UDP_IP.c(509): prvCallUDPApplicationHook
 4      70      51   FreeRTOS_UDP_IP.c(434): prvHandleUDPPacketWithoutSocket
 7     113      84   FreeRTOS_UDP_IP.c(315): vProcessGeneratedUDPPacket
 8     118      84   FreeRTOS_UDP_IP.c(574): xProcessReceivedUDPPacket
~~~
The earlier scores were like this:
~~~
27     212     166   FreeRTOS_UDP_IP.c(339): xProcessReceivedUDPPacket
39     275     208   FreeRTOS_UDP_IP.c(50): vProcessGeneratedUDPPacket
~~~

I have tested the code in the usual way:

- Look-up a host, ping it, both on the LAN and the web.
- Use and check DHCP.
- Look up an NTP server and ask ( with UDP ) for the current time.
- Repeat the above test but with IPv6 hosts.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]

File: FreeRTOS_UDP_IP.c [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
- * FreeRTOS+TCP V2.3.1
- * Copyright (C) 2020 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ * FreeRTOS+FAT V2.3.3
```
[the remark this answers is not in the document]
Comment (htibosch) [reply FreeRTOS_UDP_IP.c:2]: Indeed not intended. I copied the latest header from the +FAT repo, forgetting to replace FAT with TCP.

**Review by AniruddhaKanhere — APPROVED** [review]

File: FreeRTOS_UDP_IP.c [hunk 2]
```diff
@@ -59,28 +60,129 @@
...
+
+
 /**
- * @brief Process the generated UDP packet and do other checks before sending the
- *        packet such as ARP cache check and address resolution.
+ * @brief Look-up the target IP-address, works for both IPv4 and IPv6.
  *
- * @param[in] pxNetworkBuffer: The network buffer carrying the packet.
+ * @param[in] pxNetworkBuffer: The network buffer carrying the UDP or ICMP packet.
+ * @param[in,out] pulIPAddress: only used for IPv4: the IP-address of the destination.
```
Comment (paulbartell) [inline FreeRTOS_UDP_IP.c:88]: How is routed IPv6 handled then?

Comment (htibosch) [reply FreeRTOS_UDP_IP.c:88]: When I wrote the Doxygen comments, I notice the asymmetry: only an IPv4 address is provided! The IPv6 address is passed indirectly, as `pxUDPPacket->xEthernetHeader.xDestinationAddress`.

The IP-address pointed to by `pulIPAddress` may be replaced by the gateway.

It looks like this needs some revision, I will come back to this.

**Review by AniruddhaKanhere — APPROVED** [review]

**Review by AniruddhaKanhere — APPROVED** [review]
Just one minor comment.

**Review by aggarg — APPROVED** [review]

**Pull request merged**